src/azure-cli-core/azure/cli/core/_profile.py

```diff
@@ -174,52 +180,52 @@ def _set_subscriptions(self, new_subscriptions):
             if not new_active_one:
                 new_active_one = new_subscriptions[0]
             new_active_one[_IS_DEFAULT_SUBSCRIPTION] = True
+            default_sub_id = new_active_one[_SUBSCRIPTION_ID]
         else:
             new_subscriptions[0][_IS_DEFAULT_SUBSCRIPTION] = True
+            default_sub_id = new_subscriptions[0][_SUBSCRIPTION_ID]
 
-        self._cache_subscriptions_to_local_storage(subscriptions)
+        set_cloud_subscription(active_cloud.name, default_sub_id)
+        self._storage[_SUBSCRIPTIONS] = subscriptions
 
     def set_active_subscription(self, subscription):  # take id or name
-        subscriptions = self.load_cached_subscriptions()
-
+        subscriptions = self.load_cached_subscriptions(all_clouds=True)
+        active_cloud = get_active_cloud()
         subscription = subscription.lower()
         result = [x for x in subscriptions
-                  if subscription in [x[_SUBSCRIPTION_ID].lower(), x[_SUBSCRIPTION_NAME].lower()]]
+                  if subscription in [x[_SUBSCRIPTION_ID].lower(),
+                                      x[_SUBSCRIPTION_NAME].lower()] and
+                  x[_ENVIRONMENT_NAME] == active_cloud.name]
 
         if len(result) != 1:
-            raise CLIError('The subscription of "{}" does not exist or has more than'
-                           ' one match.'.format(subscription))
+            raise CLIError("The subscription of '{}' does not exist or has more than"
+                           " one match in cloud '{}'.".format(subscription, active_cloud.name))
 
         for s in subscriptions:
             s[_IS_DEFAULT_SUBSCRIPTION] = False
         result[0][_IS_DEFAULT_SUBSCRIPTION] = True
 
-        self._cache_subscriptions_to_local_storage(subscriptions)
+        set_cloud_subscription(active_cloud.name, result[0][_SUBSCRIPTION_ID])
+        self._storage[_SUBSCRIPTIONS] = subscriptions
 
     def logout(self, user_or_sp):
-        subscriptions = self.load_cached_subscriptions()
+        subscriptions = self.load_cached_subscriptions(all_clouds=True)
         result = [x for x in subscriptions
                   if user_or_sp.lower() == x[_USER_ENTITY][_USER_NAME].lower()]
         subscriptions = [x for x in subscriptions if x not in result]
 
-        # reset the active subscription if needed
-        result = [x for x in subscriptions if x.get(_IS_DEFAULT_SUBSCRIPTION)]
```

Comment:
If user logs out from the active subscription, no subscription will be set so they'll get a Please run 'az account set' to select active account. error message.
Once they set the account, everything is good again.

I thought that was reasonable instead of on logout, the CLI trying to automatically choose a subscription to set as active.